pkg/clusterfile/decoder.go

```diff
@@ -165,29 +165,63 @@ func checkAndFillCluster(cluster *v2.Cluster) error {
 	for _, env := range cluster.Spec.Env {
-		if strings.HasPrefix(env, common.EnvRegistryDomain) || strings.HasPrefix(env, common.EnvRegistryPort) || strings.HasPrefix(env, common.EnvRegistryURL) {
+		if strings.HasPrefix(env, common.EnvLocalRegistryDomain) ||
+			strings.HasPrefix(env, common.EnvLocalRegistryPort) ||
+			strings.HasPrefix(env, common.EnvLocalRegistryURL) ||
+			strings.HasPrefix(env, common.EnvExternalRegistryDomain) ||
+			strings.HasPrefix(env, common.EnvExternalRegistryPort) ||
+			strings.HasPrefix(env, common.EnvExternalRegistryURL) ||
+			strings.HasPrefix(env, common.EnvRegistryDomain) ||
+			strings.HasPrefix(env, common.EnvRegistryPort) ||
+			strings.HasPrefix(env, common.EnvRegistryURL) ||
+			strings.HasPrefix(env, common.EnvContainerRuntime) {
 			continue
 		}
 		newEnv = append(newEnv, env)
 	}
 	cluster.Spec.Env = newEnv
+
+	regConfig := v2.RegistryConfig{}
+	if cluster.Spec.Registry.LocalRegistry != nil {
+		regConfig = cluster.Spec.Registry.LocalRegistry.RegistryConfig
+
+		cluster.Spec.Env = append(cluster.Spec.Env, fmt.Sprintf("%s=%s", common.EnvLocalRegistryDomain, regConfig.Domain))
+		cluster.Spec.Env = append(cluster.Spec.Env, fmt.Sprintf("%s=%d", common.EnvLocalRegistryPort, regConfig.Port))
+		registryURL := net.JoinHostPort(regConfig.Domain, strconv.Itoa(regConfig.Port))
+		if regConfig.Port == 0 {
+			registryURL = regConfig.Domain
+		}
+		cluster.Spec.Env = append(cluster.Spec.Env, fmt.Sprintf("%s=%s", common.EnvLocalRegistryURL, registryURL))
+	}
+	if cluster.Spec.Registry.ExternalRegistry != nil {
+		regConfig = cluster.Spec.Registry.ExternalRegistry.RegistryConfig
+
+		cluster.Spec.Env = append(cluster.Spec.Env, fmt.Sprintf("%s=%s", common.EnvExternalRegistryDomain, regConfig.Domain))
+		cluster.Spec.Env = append(cluster.Spec.Env, fmt.Sprintf("%s=%d", common.EnvExternalRegistryPort, regConfig.Port))
+		registryURL := net.JoinHostPort(regConfig.Domain, strconv.Itoa(regConfig.Port))
+		if regConfig.Port == 0 {
+			registryURL = regConfig.Domain
+		}
+		cluster.Spec.Env = append(cluster.Spec.Env, fmt.Sprintf("%s=%s", common.EnvExternalRegistryURL, registryURL))
+	}
+
 	cluster.Spec.Env = append(cluster.Spec.Env, fmt.Sprintf("%s=%s", common.EnvRegistryDomain, regConfig.Domain))
-	cluster.Spec.Env = append(cluster.Spec.Env, fmt.Sprintf("%s=%d", common.EnvRegistryPort, regConfig.Port))
+	portStr := fmt.Sprintf("%d", regConfig.Port)
+	if regConfig.Port == 0 {
+		portStr = ""
+	}
+	cluster.Spec.Env = append(cluster.Spec.Env, fmt.Sprintf("%s=%s", common.EnvRegistryPort, portStr))
 	registryURL := net.JoinHostPort(regConfig.Domain, strconv.Itoa(regConfig.Port))
 	if regConfig.Port == 0 {
 		registryURL = regConfig.Domain
 	}
 	cluster.Spec.Env = append(cluster.Spec.Env, fmt.Sprintf("%s=%s", common.EnvRegistryURL, registryURL))
 
+	if cluster.Spec.ContainerRuntime.Type != "" {
+		cluster.Spec.Env = append(cluster.Spec.Env, fmt.Sprintf("%s=%s", common.EnvContainerRuntime, cluster.Spec.ContainerRuntime.Type))
```

Comment:
@VinceCui I think read container runtime type by clusterfile is a little complex. if user didn't specify CRI in clusterfile, What will happended? Then if user specify a wrong CRI type in clusterfile which different from clusterimage had, what will happeded?
Here is my suggestion: 
now we have `LABEL` in Kubefile, so we can get CRI info from it. When user use offical image, the CRI info will natural included. If CRI info is not specified in Kubefile, we use docker as a default CRI and give user a hint, if CRI info is wrong, we hint the default CRI info and return error for user. 
RFC @Starnop

Reply:
Thanks for the review. 

The first question, if user didn't specify CRI, the default value now is docker, but it may be changed to containerd in future.

The second question, it is no problem to specify the default container runtime by LABEL in the build phase. This does not conflict with the ability to specify the container runtime at runtime. But consider a scenario, what if the image supports multiple container runtimes? ? At this time, it is necessary to allow users to specify at runtime.

感谢review。
第一个问题，如果置为空，默认为docker，但后面可能会改成默认containerd。
第二个问题，在build阶段里LABEL指定默认的容器运行时是没有问题的，这和运行时可以指定容器运行时也是不冲突的，但是考虑一个场景，如果该镜像支持多个容器运行时呢？此时允许使用者在运行时可以指定就是必要的了。